### ISSUE TITLE
Neighbourhood Attention Implementation 

### ISSUE THREAD

**Issue by RisabBiswas** [lucidrains/vit-pytorch#243]
Can you please create an implementation of the recently released neighbourhood attention? The official code is very complicated. Your codes are always simple and very understandable. Thank you!

**Comment by lucidrains** [comment]
@RisabBiswas Hi Risab, I believe neighborhood attention was implemented with a custom CUDA kernel? I can tell you that my CUDA skills are below average at this point in time (although I am aiming to work my way up)

**Comment by lucidrains** [comment]
@RisabBiswas  you can probably get something similar with https://github.com/lucidrains/DALLE-pytorch/blob/main/dalle_pytorch/attention.py#L103 , but it will not be as performant as a custom kernel

**Comment by alihassanijr** [comment]
Hi all,
@lucidrains We recently pushed out our kernels as a pip package ([NATTEN](https://shi-labs.com/natten)), so NAT and DiNAT can be easily added just by making `natten` required.
I would add that we're releasing a new version that also has CPU kernels, so effectively most users should be able to run NA-based models with just the pip install.
I'd be more than happy to open a PR for those.

P.S. @RisabBiswas the reason why we've had to do all of this is because NA is a pixel-wise operation, therefore writing it with DL libraries will just end up being a really slow implementation with a really high memory footprint. We actually did have a "pure torch" implementation of NA, which has been deprecated in favor of our upcoming CPU kernels.
Basically you'd have to run an explicit loop (in Python) through all of your pixels -- really slow.
Or you'd use "sliding window" operations like `nn.Unfold` and extract a KxK window, and that's per pixel.
This is not only slow, it'll use up a lot of memory in the process.
Just to give an idea, with a highly tuned pytorch implementation, it took us **about a week to train a Tiny variant**, whereas with the kernel, now, it'll take **about a day** on the same set up (8xA100s, DDP, AMP).
Memory usage also used to grow exponentially when scaling the model, now it actually uses less memory than counterparts like Swin, and it runs up to 40% faster (all mentioned in the paper and [NATTEN GitHub](https://github.com/SHI-Labs/NATTEN)).

**Comment by lucidrains** [comment]
@alihassanijr very nice :) congrats on the paper!

any plans to make a 3d version? (for video)

**Comment by alihassanijr** [comment]
Thank you. Let me know if you need a hand adding NAT/DiNAT.
And yes, our lab is working on 3D NA for videos, but with quite limited bandwidth due to other projects.
We are very open to collaboration if you happen to be interested and have time. :slightly_smiling_face:
